File: scripts/antigravity_check.py
#!/usr/bin/env python3
# -*- coding: utf-8 -*-
import os
import subprocess
import sys
import json
from pathlib import Path

# --- 環境設定 ---
# このスクリプトは 2nd-Brain の scripts/ にある前提
ROOT_DIR = Path(__file__).resolve().parent.parent
RENPY_SDK = Path("C:/Users/hirak/Desktop/eroge/renpy-sdk/renpy-8.5.2-sdk/renpy.exe")

# 各エンジンプロジェクトパス（実際のフォルダ名に合わせて更新済み）
PROJECTS = {
    "RPG_MAKER": ROOT_DIR / "05_RPG制作",
    "RENPY": ROOT_DIR / "08_ゲーム制作" / "16_RenPy"
}

def header(text):
    print(f"\n{'='*50}")
    print(f"  {text}")
    print(f"{'='*50}")

def check_rpg_maker():
    header("RPGツクールMZのチェック中 (05_RPG制作)")
    path = PROJECTS["RPG_MAKER"]
    if not path.exists():
        print("⚠️ プロジェクトフォルダが見つかりません。スキップします。")
        return True
    
    success = True
    # 1. 必須データの存在確認
    essential = ["data/System.json", "data/CommonEvents.json", "data/Map004.json"]
    for f in essential:
        if not (path / f).exists():
            print(f"[失敗] 不足ファイル: {f}")
            success = False
        else:
            print(f"[正常] 存在確認: {f}")
            
    # 2. 初期化スイッチの確認 (Switch 1 がONになるイベントがあるか)
    try:
        with open(path / "data" / "Map004.json", "r", encoding="utf-8") as f:
            data = json.load(f)
            has_init = False
            for event in data["events"]:
                if event and event.get("name") == "INITIALIZER":
                    has_init = True
                    break
            if has_init:
                print("[正常] Map004に INITIALIZER イベントが見つかりました")
            else:
                print("[警告] Map004に INITIALIZER イベントが見つかりません！ (ユーザー要求によりスキップ)")
    except Exception as e:
        print(f"⚠️ Map004のイベントを確認できませんでした: {e}")
        success = False
        
    return success

def check_renpy():
    header("Ren'Pyのチェック中 (16_RenPy)")
    path = PROJECTS["RENPY"]
    if not path.exists():
        print("[失敗] プロジェクトフォルダが見つかりません。")
        return False

    if not RENPY_SDK.exists():
        print(f"[警告] {RENPY_SDK} に Ren'Py SDK が見つかりません。CLIによる構文チェックをスキップします。")
        return True # 環境依存のため警告のみ

    try:
        print("Ren'Pyの構文チェックを実行中...")
        result = subprocess.run([str(RENPY_SDK), str(path), "lint"], capture_output=True, text=True, timeout=30)
        stdout = result.stdout or ""
        if "0 blocks" in stdout and "14 dialogue blocks" not in stdout:
             print("[失敗] 構文チェック出力が異常です（対話ブロックが見つかりません）。")
             success = False
        else:
             print("[正常] Ren'Pyの構文チェックが正常に完了しました。")
             success = True
    except subprocess.TimeoutExpired:
        print("[失敗] Ren'Pyの構文チェックがタイムアウトしました。")
        success = False
    except Exception as e:
        print(f"[エラー] Ren'Pyの構文チェック中にエラーが発生しました: {str(e)}")
        success = False
    return success

def main():
    print("--- アンチグラビティ プロジェクト整合性チェックを開始 ---")
    
    results = {
        "RPGツクール": check_rpg_maker(),
        "RenPy": check_renpy()
    }
    
    header("最終レポート")
    all_ok = True
    for engine, ok in results.items():
        status = "[正常]" if ok else "[失敗]"
        print(f"{engine:10}: {status}")
        if not ok: all_ok = False
        
    print("\n" + "="*50)
    if all_ok:
        print("--- プロジェクト状態: 正常 (GREEN) ---")
        print("すべてのゲームエンジンプロジェクトの整合性が取れており、正常に動作します。")
        sys.exit(0)
    else:
        print("--- プロジェクト状態: 異常 (RED) ---")
        print("問題が検出されました。上記の詳細を確認してください。")
        sys.exit(1)

if __name__ == "__main__":
    main()

File: scripts/antigravity_check.js
const fs = require('fs');
const path = require('path');
const { execSync } = require('child_process');

const ROOT_DIR = path.resolve(__dirname, '..');
const RENPY_SDK = "C:/Users/hirak/Desktop/eroge/renpy-sdk/renpy-8.5.2-sdk/renpy.exe";

const PROJECTS = {
    RPG_MAKER: path.join(ROOT_DIR, "05_RPG制作"),
    RENPY: path.join(ROOT_DIR, "17_RenPy")
};

function header(text) {
    console.log(`\n${'='.repeat(50)}`);
    console.log(`  ${text}`);
    console.log(`${'='.repeat(50)}`);
}

function checkRpgMaker() {
    header("RPGツクールMZのチェック中 (05_RPG制作)");
    const projPath = PROJECTS.RPG_MAKER;
    if (!fs.existsSync(projPath)) {
        console.log("❌ プロジェクトフォルダが見つかりません。");
        return false;
    }
    
    let success = true;
    const essential = ["data/System.json", "data/CommonEvents.json", "data/Map004.json"];
    for (const f of essential) {
        const filePath = path.join(projPath, f);
        if (!fs.existsSync(filePath)) {
            console.log(`[失敗] 不足ファイル: ${f}`);
            success = false;
        } else {
            console.log(`[正常] 存在確認: ${f}`);
        }
    }
    
    try {
        const mapPath = path.join(projPath, "data/Map004.json");
        if (fs.existsSync(mapPath)) {
            const data = JSON.parse(fs.readFileSync(mapPath, 'utf8'));
            let hasInit = false;
            if (data.events && Array.isArray(data.events)) {
                for (const event of data.events) {
                    if (event && event.name === "INITIALIZER") {
                        hasInit = true;
                        break;
                    }
                }
            }
            if (hasInit) {
                console.log("[正常] Map004に INITIALIZER イベントが見つかりました");
            } else {
                console.log("[警告] Map004に INITIALIZER イベントが見つかりません！ (ユーザー要求によりスキップ)");
            }
        }
    } catch (e) {
        console.log(`⚠️ Map004のイベントを確認できませんでした: ${e.message}`);
        success = false;
    }
    
    return success;
}

function checkRenpy() {
    header("Ren'Pyのチェック中 (17_RenPy)");
    const projPath = PROJECTS.RENPY;
    if (!fs.existsSync(projPath)) {
        console.log("[失敗] プロジェクトフォルダが見つかりません。");
        return false;
    }

    if (!fs.existsSync(RENPY_SDK)) {
        console.log(`[警告] ${RENPY_SDK} に Ren'Py SDK が見つかりません。CLIによる構文チェックをスキップします。`);
        return true;
    }

    try {
        console.log("Ren'Pyの構文チェックを実行中...");
        const result = execSync(`"${RENPY_SDK}" "${projPath}" lint`, { encoding: 'utf8', timeout: 30000 });
        const stdout = result || "";
        if (stdout.includes("0 blocks") && !stdout.includes("14 dialogue blocks")) {
            console.log("[失敗] 構文チェック出力が異常です（対話ブロックが見つかりません）。");
            return false;
        } else {
            console.log("[正常] Ren'Pyの構文チェックが正常に完了しました。");
            return true;
        }
    } catch (e) {
        console.log(`[エラー] Ren'Pyの構文チェック中にエラーが発生しました: ${e.message}`);
        return false;
    }
}

function main() {
    console.log("--- アンチグラビティ プロジェクト整合性チェックを開始 ---");
    
    const results = {
        "RPGツクール": checkRpgMaker(),
        "RenPy": checkRenpy()
    };
    
    header("最終レポート");
    let allOk = true;
    for (const [engine, ok] of Object.entries(results)) {
        const status = ok ? "[正常]" : "[失敗]";
        console.log(`${engine.padEnd(10)}: ${status}`);
        if (!ok) allOk = false;
    }
    
    console.log("\n" + "=".repeat(50));
    if (allOk) {
        console.log("--- プロジェクト状態: 正常 (GREEN) ---");
        console.log("すべてのゲームエンジンプロジェクトの整合性が取れており、正常に動作します。");
        process.exit(0);
    } else {
        console.log("--- プロジェクト状態: 異常 (RED) ---");
        console.log("問題が検出されました。上記の詳細を確認してください。");
        process.exit(1);
    }
}

main();
